src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +25,649 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+            cachedInfos.Add(declarationInfo);
 
-            // Must contain equivalents named types residing in different assemblies.
-            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => !SymbolEquivalenceComparer.Instance.Equals(kvp.Key.ContainingAssembly, kvp.Value.ContainingAssembly)));
+            HashSet<INamedTypeSymbol> result = null;
+            foreach (var symbolInfo in declarationInfo.DeclaredSymbolInfos)
+            {
+                result = await ProcessSymbolInfo(
+                    document, symbolInfo,
+                    typesToSearchFor,
+                    inheritanceInfo, cachedModels,
+                    typeImmediatelyMatches, result, cancellationToken).ConfigureAwait(false);
+            }
 
-            // Must contain non-nested named types.
-            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Key.ContainingType == null));
-            Contract.ThrowIfFalse(equivalentTypesWithDifferingAssemblies.All(kvp => kvp.Value.ContainingType == null));
+            return result;
+        }
 
-            var referencedAssemblies = new MultiDictionary<string, IAssemblySymbol>();
-            foreach (var assembly in compilation.GetReferencedAssemblySymbols())
+        private static async Task<HashSet<INamedTypeSymbol>> ProcessSymbolInfo(
+            Document document,
+            DeclaredSymbolInfo info,
+            HashSet<INamedTypeSymbol> typesToSearchFor,
+            InheritanceInfo inheritanceInfo,
+            ConcurrentSet<SemanticModel> cachedModels,
+            Func<HashSet<INamedTypeSymbol>, INamedTypeSymbol, bool> typeImmediatelyMatches,
+            HashSet<INamedTypeSymbol> result,
+            CancellationToken cancellationToken)
+        {
+            // If we're searching for enums/structs/delegates, then we can just look at the kind of
+            // the info to see if we have a match.
+            if ((inheritanceInfo.DerivesFromSystemEnum && info.Kind == DeclaredSymbolInfoKind.Enum) ||
+                (inheritanceInfo.DerivesFromSystemValueType && info.Kind == DeclaredSymbolInfoKind.Struct) ||
+                (inheritanceInfo.DerivesFromSystemMulticastDelegate && info.Kind == DeclaredSymbolInfoKind.Delegate))
             {
-                referencedAssemblies.Add(assembly.Name, assembly);
+                var symbol = await ResolveAsync(document, info, cachedModels, cancellationToken).ConfigureAwait(false) as INamedTypeSymbol;
+                if (symbol != null)
+                {
+                    result = result ?? new HashSet<INamedTypeSymbol>();
+                    result.Add(symbol);
+                }
             }
-
-            int verifiedCount = 0;
-            foreach (var kvp in equivalentTypesWithDifferingAssemblies)
+            else if (inheritanceInfo.DerivesFromSystemObject && info.Kind == DeclaredSymbolInfoKind.Class)
```

Comment:
The implication by this code is inheritenceInfo can lie? Should it be named accordingly?